packages/indexer/src/starknet/parser.ts

```diff
@@ -0,0 +1,56 @@
+import { FieldElement } from "./felt";
+import { EventFilter } from "./filter";
+import { EventWithTransaction } from "./block";
+
+import { type Abi, hash } from "starknet";
+
+interface EventFilterParser {
+  eventFilter(name: string): EventFilter;
+  eventParser(name: string): ({ event, transaction }: EventWithTransaction ) => any;
+}
+
+// The Contract class implements the EventFilterParser interface.
+export class Contract implements EventFilterParser {
+  // Read-only properties for the contract's address and its ABI.
+  readonly contractAddress: FieldElement;
+  readonly contractAbi: Abi;
+
+  // Constructor to initialize a Contract instance with a contract address and ABI.
+  constructor(contractAddress: FieldElement, contractAbi: Abi) {
+    this.contractAddress = contractAddress;
+    this.contractAbi = contractAbi;
+  }
+
+  // Method to filter events based on their name from the contract's ABI.
+  eventFilter(name: string): EventFilter {
+    // Find the event in the ABI matching the provided name.
+    const event: EventFilter = this.contractAbi.find((item) => item.name === name);
+    
+    // Throw an error if the event is not found in the ABI
+    if (!event) {
+      throw new Error(`Event ${name} not found in contract ABI`);
+    }
+
+    // Return the found event.
+    return event;
+  }
+
+  // Method to parse events, given their name, and return a function that processes 
+  // an event and its associated transaction.
+  eventParser(name: string): ({ event, transaction }: EventWithTransaction ) => any {
+    const event = this.eventFilter(name);
+
+    // Need clarity on how to get a transaction from an event.
```

Comment:
I think the goal of this function will become much clearer if you play around with Apibara. I recommend installing the CLI tool like in the getting started guide I linked, then run the following script with `apibara run script.js -A dna_<xxx>` (`dna_<xx>` is your API key, you need an account).


```ts
import { hash } from "https://esm.run/starknet";

export const config = {
  streamUrl: "https://mainnet.starknet.a5a.ch",
  startingBlock: 100_000,
  network: "starknet",
  batchSize: 1,
  finality: "DATA_STATUS_PENDING",
  filter: {
    header: { weak: true },
    events: [
      {
        fromAddress:
          "0x053c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8",
        keys: [hash.getSelectorFromName("Transfer")],
        includeTransaction: false,
        includeReceipt: false,
      },
    ],
  },
  sinkType: "console",
  sinkOptions: {},
};

export default function transform({ header, events }) {
  const { blockNumber, timestamp } = header;
  return events.map(({ event }) => ({
    ...event,
    blockNumber,
    timestamp,
  }));
}
```

The goal of `eventParser` is to convert the "raw" event to an higher-level representation.

```ts
const parser = contract.eventParser("Transfer");

export default function transform({ header, events }) {
  const { blockNumber, timestamp } = header;
  return events.map(parser);
}
```

Reply:
We will use starknet.js to do the hard job of parsing the event, so don't worry!